lib/lte_link_control/lte_lc.c

```diff
@@ -598,64 +623,82 @@ static int connect_lte(bool blocking)
 		goto exit;
 	}
 
-	do {
-		tries--;
+	err = lte_lc_func_mode_get(&original_func_mode);
+	if (err) {
+		err = -EFAULT;
+		goto exit;
+	}
+
+	if (lte_lc_sys_mode != LTE_LC_SYSTEM_MODE_DEFAULT &&
+	    IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK)) {
+		/* Set system mode, because it may have been changed by fallback. */
+		if (!(original_func_mode == LTE_LC_FUNC_MODE_POWER_OFF ||
+		      original_func_mode == LTE_LC_FUNC_MODE_OFFLINE)) {
+			err = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_OFFLINE);
+			if (err) {
+				err = -EFAULT;
+				goto exit;
+			}
+		}
 
-		err = lte_lc_func_mode_get(&current_func_mode);
+		err = lte_lc_system_mode_set(lte_lc_sys_mode, lte_lc_sys_mode_pref);
 		if (err) {
 			err = -EFAULT;
 			goto exit;
 		}
+	}
 
-		/* Change the modem sys-mode only if it's not running or is meant to change */
-		if (!IS_ENABLED(CONFIG_LTE_NETWORK_MODE_DEFAULT) &&
-		    ((current_func_mode == LTE_LC_FUNC_MODE_POWER_OFF) ||
-		     (current_func_mode == LTE_LC_FUNC_MODE_OFFLINE))) {
-			err = lte_lc_system_mode_set(lte_lc_sys_mode, lte_lc_sys_mode_pref);
+	/* Reset the semaphore, it may have already been given
+	 * by an earlier +CEREG notification.
+	 */
+	k_sem_reset(&link);
+
+	err = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_NORMAL);
+	if (err || !blocking) {
+		goto exit;
+	}
+
+	err = k_sem_take(&link, K_SECONDS(CONFIG_LTE_NETWORK_TIMEOUT));
+	if (err == -EAGAIN) {
+		LOG_INF("Network connection attempt timed out");
+		err = -ETIMEDOUT;
+
+		if (IS_ENABLED(CONFIG_LTE_NETWORK_USE_FALLBACK)) {
+			err = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_OFFLINE);
 			if (err) {
 				err = -EFAULT;
 				goto exit;
 			}
-		}
 
-		/* Reset the semaphore, it may have already been given
-		 * by an earlier +CEREG notification.
-		 */
-		k_sem_reset(&link);
+			err = fallback_system_mode_set();
+			if (err) {
+				err = -ETIMEDOUT;
+				goto exit;
+			}
 
-		err = lte_lc_func_mode_set(LTE_LC_FUNC_MODE_NORMAL);
-		if (err || !blocking) {
-			goto exit;
-		}
+			LOG_INF("Using fallback system mode");
```

Comment:
If you include this info into log on line 590, maybe this could be removed.

Reply:
Yes, I'll remove this.